### PR TITLE
Add na.rm to aggregating functions

### PR DESCRIPTION
**Pull request by TuomasBorman**
This PR adds `na.rm` option to following functions: `sumCountsAcrossFeatures`, `numDetectedAcrossFeatures`, `summarizeAssayByGroup`, `numDetectedAcrossCells`.

This PR is related to this issue: https://github.com/LTLA/scuttle/issues/33

### CODE REVIEW

**Comment by LTLA** [comment]
Thanks for the PR. I'll address some immediate points first, but having slept on your problem, I would recommend a much simpler solution.

- `r == NA_INTEGER` is not correct if the matrix was double-precision, as the `NA_INTEGER` value (-2^31) is a valid double-precision number. You'll need to check the input matrix type and only perform that check if the input matrix is integer or logical.
- Conversely, `std::isnan()` is unnecessary when the input matrix is integer or logical, as IEEE specials wouldn't fit inside a 32-bit integer anyway.
- All of the new checks are unnecessary when `na.rm=FALSE`. To avoid having a potential optimization-breaking `if` statement in the tight inner `while` loop, hoist it out and duplicate the `for (auto s : runs)` loop for each NA'ness status (`na.rm=TRUE` with double, `na.rm=TRUE` with int, `na.rm=FALSE`) with only the relevant NA checks in each one.
- You have too many comments - don't touch the lines that you don't need to.
- No pipes in my packages.

You can see that it is quite tedious to protect against `NA`s while preserving performance for the vast majority of (single-cell) use cases that don't need it. So perhaps you might consider a much simpler approach:

```r
y <- rsparsematrix(10000, 5, density=0.1)
z <- sample(LETTERS[1:5], nrow(y), replace=TRUE)
y[1] <- NA

y <- DelayedArray(y)
is.okay <- !is.na(y)
type(is.okay) <- "double"
DelayedArray::rowsum(y, z, na.rm=TRUE) / DelayedArray::rowsum(is.okay, z)
``` 

4 lines and done. No copying involved as the `DelayedArray()` defers the `is.na` and type coercion operations. Maybe a bit slower than the C++ code but I doubt it matters much.

**Comment by TuomasBorman** [comment]
Thank you for the feedback! I'm not very experienced with C++, so I jumped straight into the deep end. Your suggestions seems to be working and it is much simpler. I would still tend to think that scuttle is the right place to fix this `na.rm` issue. Would it be ok, if I

1. revert changes affecting the C++ code
2. add `na.rm` to `.sum_across_features`: if `na.rm`, then run _your 4 lines_ instead of C++ code

Moreover, `summarizeAssayByGroup` used loops (`lapply`) to aggregate data. I guess this approach could also improve the efficiency of that function.

**Comment by LTLA** [comment]
I had a long hard think about this last night. Your suggestion is pragmatic, but the problem is that the rest of **scuttle** doesn't know or care about `NA`s in the data matrix. Accepting this change would actually represent an expansion of **scuttle**'s scope to handle `NA`s. At that point, I would need to modify many **scuttle** functions to be `NA`-safe, because it would be very confusing for some functions to be `NA`-aware and others to be `NA`-unaware. And there are a _lot_ of functions...

I think that the heart of the issue is the purpose of **scuttle**. For its intended purpose of single-cell data analysis, `NA`s never show up, even after merging datasets. (I typically take the intersection of features across datasets for downstream analyses to avoid having to deal with `NA`s.) Whether **scuttle** can be used in other applications depends on how well-aligned those applications are to the single-cell context, and in the case of `NA`s, it appears there are some differences.

In short, I don't think I can accept your suggestion. I was a bit too thoughtless in suggesting that you make a PR without considering the implications - sorry about that. I hope you got some learning experiences re. C++ at least. Perhaps this topic can be revisited in the future if **scuttle** needs to be made `NA`-aware, which would be a major project in itself.

In the meantime, the `DelayedArray::rowsum()` approach is what I would recommend.

**Comment by TuomasBorman** [comment]
Alright, I understand your point. `DelayedArray::rowsum()`-approach was neater than I came up in mia package so this discussion and exploration with different approaches were helpful for me.
